src/core/PluginPinConnector.cpp

```diff
@@ -0,0 +1,527 @@
+/*
+ * PluginPinConnector.cpp - Specifies how to route audio channels
+ *                          in and out of a plugin.
+ *
+ * Copyright (c) 2024 Dalton Messmer <messmer.dalton/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "PluginPinConnector.h"
+
+#include <QDomDocument>
+#include <QDomElement>
+#include <cassert>
+#include <stdexcept>
+
+#include "AudioEngine.h"
+#include "Model.h"
+#include "PluginPinConnectorView.h"
+
+namespace lmms
+{
+
+PluginPinConnector::PluginPinConnector(Model* parent)
+	: Model{parent}
+{
+	updateTrackChannels(s_totalTrackChannels);
+}
+
+PluginPinConnector::PluginPinConnector(int pluginInCount, int pluginOutCount, Model* parent)
+	: Model{parent}
+{
+	updateTrackChannels(s_totalTrackChannels);
+	setChannelCounts(pluginInCount, pluginOutCount);
+}
+
+auto PluginPinConnector::Matrix::channelName(int channel) const -> QString
+{
+	// Custom name (if supported)
+	assert(channel >= 0);
+	if (channel < static_cast<int>(channelNames.size()))
+	{
+		return channelNames[channel];
+	}
+
+	// A-Z
+	if (channel < 26)
+	{
+		return QChar::fromLatin1('A' + channel);
+	}
+
+	// AA-ZZ
+	channel -= 26;
+	if (channel < 26 * 26)
+	{
+		auto ret = QString{"AA"};
+		ret[0] = QChar::fromLatin1('A' + channel / 26);
+		ret[1] = QChar::fromLatin1('A' + channel % 26);
+		return ret;
+	}
+
+	throw std::invalid_argument{"Too many channels"};
+}
+
+void PluginPinConnector::setChannelCounts(int inCount, int outCount)
+{
+	if (m_trackChannelsUsed > MaxTrackChannels)
+	{
+		throw std::runtime_error{"Only up to 256 track channels are allowed"};
+	}
+
+	if (inCount < 0)
+	{
+		throw std::invalid_argument{"Invalid input count"};
```

Comment:
I think it's good here since `setPluginChannelCountIn` calls this method, and I want the check to still be there when this method is called on its own.